File: javascript/crypto-chunk-cipher/cryptoChunkCipher.js
/*
 * (c) 2018-2020 Ionic Security Inc.
 * By using this code, I agree to the Terms & Conditions (https://dev.ionic.com/use.html)
 * and the Privacy Policy (https://www.ionic.com/privacy-notice/).
 */

/*
 * WARNING *
 * Calling agent.enrollUser() successfully is a prerequisite before using this code.
 * This is done using enrollDevice.js.
 */

import {getAgentConfig} from '../jssdkConfig.js';

const main = async () => {
  const message = 'secret message'

  const appData = getAgentConfig('Javascript Crypto Chunk Cipher');

  // initialize agent
  try {
    const resp = await new window.IonicSdk.ISAgent(appData);
    const agent = resp.agent;

    // encrypt message
    let ciphertext;
    try {
      const encryptResponse = await agent.encryptStringChunkCipher({stringData: message})
      ciphertext = encryptResponse.stringChunk
    } catch (SdkErrorResponse) {
      console.log('Encryption error: ' + SdkErrorResponse.error);
    }

    // decrypt message
    try {
      const decryptResponse = await agent.decryptStringChunkCipher({stringData: ciphertext})
      const plaintext = decryptResponse.stringChunk

      // display data
      console.log('');
      console.log('Ciphertext : ' + ciphertext)
      console.log('Plaintext  : ' + plaintext)
    } catch (SdkErrorResponse) {
      console.log('Decryption error: ' + SdkErrorResponse.error);
    }
  } catch (sdkErrorResponse) {
    console.log('Initializing agent error: ' + sdkErrorResponse.error);
  }
}

main()

File: cpp/crypto-chunk-cipher/src/CryptoChunkCipher.cpp
/*
 * (c) 2018-2020 Ionic Security Inc.
 * By using this code, I agree to the Terms & Conditions (https://dev.ionic.com/use.html)
 * and the Privacy Policy (https://www.ionic.com/privacy-notice/).
 */

#include "ISAgent.h"
#include "ISAgentSDKError.h"
#include <ISChunkCrypto.h>
#include <stdio.h>
#include <cstdlib>
#include <iostream>

#ifdef _WIN32
    #define HOMEVAR "USERPROFILE"
#else 
    #define HOMEVAR "HOME"
#endif

int main(int argc, char* argv[]) {

    int nErrorCode;
    std::string message = "secret message";

    // read persistor password from environment variable
    char* cpersistorPassword = std::getenv("IONIC_PERSISTOR_PASSWORD");
    if (cpersistorPassword == NULL) {
        std::cerr << "[!] Please provide the persistor password as env variable: IONIC_PERSISTOR_PASSWORD" << std::endl;
        exit(1);
    }
    std::string persistorPassword = std::string(cpersistorPassword);

    // initialize agent with password persistor
    std::string persistorPath = std::string(std::getenv(HOMEVAR)) + "/.ionicsecurity/profiles.pw";
    ISAgentDeviceProfilePersistorPassword persistor;
    persistor.setFilePath(persistorPath);
    persistor.setPassword(persistorPassword);
    ISAgent agent;
    nErrorCode = agent.initialize(persistor);
    if (nErrorCode != ISAGENT_OK) {
        std::cerr << "Failed to initialize agent from password persistor (" << persistorPath << ")" << std::endl;
        std::cerr << ISAgentSDKError::getErrorCodeString(nErrorCode) << std::endl;
        exit(1);
    }

    // initialize chunk cipher object
    ISChunkCryptoCipherAuto cipher(agent);

    // encrypt
    std::string ciphertext;
    nErrorCode = cipher.encrypt(message, ciphertext);
    if (nErrorCode != ISCRYPTO_OK) {
        std::cerr << "Error: " << ISAgentSDKError::getErrorCodeString(nErrorCode) << std::endl;
        exit(1);
    }

    // decrypt
    std::string plaintext;
    cipher.decrypt(ciphertext, plaintext);
    if (nErrorCode != ISCRYPTO_OK) {
        std::cerr << "Error: " << ISAgentSDKError::getErrorCodeString(nErrorCode) << std::endl;
        exit(1);
    }

    // display data
    std::cout << "Ciphertext : " << ciphertext << std::endl;
    std::cout << "Plaintext  : " << plaintext << std::endl;
}

